lumeer-core/src/main/java/io/lumeer/core/task/executor/AutoLinkRuleTaskExecutor.java

```diff
@@ -0,0 +1,175 @@
+      this.ruleName = ruleName;
+      this.rule = new AutoLinkRule(ruleTask.getRule());
+      this.ruleTask = ruleTask;
+   }
+
+   public void execute() {
+      final LinkType linkType = ruleTask.getDaoContextSnapshot().getLinkTypeDao().getLinkType(rule.getLinkType());
+
+      if (linkType != null) {
+         final Document thisDocument = ruleTask.getOldDocument() != null ? ruleTask.getOldDocument() : ruleTask.getNewDocument();
+         final String thisCollectionId = thisDocument.getCollectionId();
+         final String thatCollectionId = linkType.getCollectionIds().get(0).equals(thisCollectionId) ? linkType.getCollectionIds().get(1) : linkType.getCollectionIds().get(0);
+         final String thisAttribute, thatAttribute;
+
+         if (rule.getCollection1().equals(thisCollectionId)) {
+            thisAttribute = rule.getAttribute1();
+            thatAttribute = rule.getAttribute2();
+         } else {
+            thisAttribute = rule.getAttribute2();
+            thatAttribute = rule.getAttribute1();
+         }
+
+         // both documents are set
+         if (ruleTask.getOldDocument() != null && ruleTask.getNewDocument() != null) {
+            Object o1 = ruleTask.getOldDocument().getData().get(thisAttribute);
+            Object o2 = ruleTask.getNewDocument().getData().get(thisAttribute);
+
+            // the attributes are different
+            if ((o1 != null && !o1.equals(o2)) || (o2 != null && !o2.equals(o1))) {
+               // it was not null before
+               if (o1 != null) {
+                  removeLinks(ruleTask.getOldDocument(), linkType, thatCollectionId, thisAttribute, thatAttribute);
+               }
+               // and it is not null either
+               if (o2 != null) {
+                  addLinks(ruleTask.getNewDocument(), linkType, thatCollectionId, thisAttribute, thatAttribute);
+               }
+            }
+         } else { // one of the docs is null (i.e. new document created or old document deleted
+            if (ruleTask.getOldDocument() != null && ruleTask.getOldDocument().getData().get(thisAttribute) != null) {
+               removeLinks(ruleTask.getOldDocument(), linkType, thatCollectionId, thisAttribute, thatAttribute);
+            }
+            if (ruleTask.getNewDocument() != null && ruleTask.getNewDocument().getData().get(thisAttribute) != null) {
+               addLinks(ruleTask.getNewDocument(), linkType, thatCollectionId, thisAttribute, thatAttribute);
+            }
+         }
+      }
+   }
+
+   private void removeLinks(final Document oldDocument, final LinkType linkType, final String thatCollection, final String thisAttribute, final String thatAttribute) {
+      final String thisCollection = oldDocument.getCollectionId();
+
+      final SearchQuery query = SearchQuery
+            .createBuilder()
+            .stems(Arrays.asList(
+                  SearchQueryStem
+                        .createBuilder("")
```

Comment:
`oldDocument.getCollectionId()`

Reply:
There is no reason to set collection id, it is not used anywhere. There just isn't any other constructor. IMHO Query and QueryStem are not good for Dao layer.